torch/ao/quantization/pt2e/duplicate_dq_pass.py

```diff
@@ -0,0 +1,76 @@
+import logging
+
+import torch
+from torch._export.pass_base import _ExportPassBase
+
+from torch.ao.quantization.pt2e.utils import (
+    _filter_sym_size_users,
+    _is_valid_annotation,
+)
+
+from torch.fx.node import map_arg
+from torch.fx.passes.infra.pass_base import PassResult
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.WARNING)
+
+__all__ = ["DuplicateDQPass"]
+
+_DEQUANTIZE_OPS = [
+    torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+    torch.ops.quantized_decomposed.dequantize_per_tensor.tensor,
+    torch.ops.quantized_decomposed.dequantize_per_channel.default,
+]
+
+
+"""
+This API is specifically to replace a specific use of producer node
+ATM, it uses internal API __update_args_kwargs, but ideally we mvoe
+this API to fx.Node.replace_use_with...
+"""
+
+
+def _copy_and_replace_use(
+    gm: torch.fx.GraphModule, producer: torch.fx.Node, user: torch.fx.Node
+):
+    with gm.graph.inserting_after(producer):
+        new_node = gm.graph.node_copy(producer)
+
+        def maybe_replace_node(n: torch.fx.Node) -> torch.fx.Node:
+            if n == producer:
+                return new_node
+            else:
+                return n
+
+        new_args = map_arg(user.args, maybe_replace_node)
+        new_kwargs = map_arg(user.kwargs, maybe_replace_node)
+        user.args = new_args
+        user.kwargs = new_kwargs
+
+
+def _maybe_duplicate_dq(
+    gm: torch.fx.GraphModule, producer: torch.fx.Node, user: torch.fx.Node
+):
+    annotation = user.meta.get("quantization_annotation", None)
+    if not _is_valid_annotation(annotation):
+        return
+    _copy_and_replace_use(gm, producer, user)
+
+
+class DuplicateDQPass(_ExportPassBase):
```

Comment:
I'm wondering if it makes sense to just always duplicate dequant for now, or do a duplicate and then deduplicate like we have did before in https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/pt2e/qat_utils.py#L580-L627

Reply:
Oh you mean, move this before fold_conv_bn? If so yes. makes sense.

Reply:
I was wondering if we can do:
```
duplicate all dequant -> port metadata -> deduplicate dequantize ops when we didn't find metadatas on them
```

Reply:
why is that better? It might be cleaner if we dont duplicate and dedup.

Reply:
so current quantization_annotation doesn't really convey any information about share or not share dq nodes, I think it will be better to not rely on it

actually I'm wondering when does this have to be landed? I'm thinking of doing a proper design for this part next week, right now this information is missing from QuantizationSpec

Reply:
there is no sharing of dq nodes at all. The only reason annotation is taken into account is so that you dont unnecessarily duplicate dq nodes. I can very much remove quantization_annotation from decision making, it will just have performance imlication